lib/hg-repository-provider.coffee

```diff
@@ -0,0 +1,41 @@
+HgRepository = require './hg-repository'
+
+findRepositoryRoot = (directory) ->
+  hgDir = directory.getSubdirectory('.hg')
+  if hgDir.existsSync?()
+    return directory
+  else if directory.isRoot()
+    return null
+  else
+    findRepositoryRoot(directory.getParent())
+
+module.exports =
+  class HgRepositoryProvider
+    constructor: (@project) ->
+      @pathToRepository = {}
+
+    repositoryForDirectory: (directory) ->
+      Promise.resolve(@repositoryForDirectorySync(directory))
+
+    repositoryForDirectorySync: (directory) ->
+      repositoryRoot = findRepositoryRoot(directory)
+      unless repositoryRoot
+        return null
+
+      repositoryPath = repositoryRoot.getPath()
+      if !@pathToRepository
+        @pathToRepository = {}
+
+      repo = @pathToRepository[repositoryPath]
+      unless repo
+        repo = HgRepository.open(repositoryPath, project: @project)
+        return null unless repo
+
+        # TODO: takes first repository only
```

Comment:
I'll merge this pull request. If you answer yes to the question above I'll also remove this TODO later.